src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.Snapshotter.cs

```diff
@@ -0,0 +1,250 @@
+                        {
+                            break;
+                        }
+
+                        variableBySlotBuilder.Add(walkerVariableBySlot[i]);
+                    }
+
+#if DEBUG
+                    verifyRemainderNonExistant(walkerVariableBySlot, i);
+#endif
+
+                    _lastVariableBySlot = variableBySlotBuilder.ToImmutableAndFree();
+                    _lastVariableSlot = _walkerToSnapshot._variableSlot.ToImmutableDictionary();
+                }
+#if DEBUG
+                else
+                {
+                    // In debug mode, we verify the contents of the arrays to ensure that the invariant
+                    // is maintained. Because variableBySlot is allocated exponentially, we can only
+                    // verify that the length of our cached version is equal or less than the source
+                    // array, and then verify that anything we haven't cached does not exist.
+                    Debug.Assert(_lastVariableBySlot.Length <= _walkerToSnapshot.variableBySlot.Length);
+                    int i;
+                    for (i = 0; i < _lastVariableBySlot.Length; i++)
+                    {
+                        Debug.Assert(compareIdentifiers(_lastVariableBySlot[i], _walkerToSnapshot.variableBySlot[i]));
+                    }
+                    verifyRemainderNonExistant(_walkerToSnapshot.variableBySlot, i);
+
+                    Debug.Assert(_lastVariableSlot.Count == _walkerToSnapshot._variableSlot.Count);
+                    foreach ((VariableIdentifier key, int value) in _lastVariableSlot)
+                    {
+                        if (!key.Exists) continue;
+                        Debug.Assert(_walkerToSnapshot._variableSlot.ContainsKey(key));
+                        Debug.Assert(value == _walkerToSnapshot._variableSlot[key]);
+                    }
+                }
+#endif
+
+                return new VariableState(
+                    _lastVariableSlot,
+                    _lastVariableBySlot,
+                    _walkerToSnapshot._variableTypes,
+                    localState.HasValue ? localState.Value : _walkerToSnapshot.State.Clone());
+
+                static bool compareIdentifiers(VariableIdentifier id1, VariableIdentifier id2) =>
+                    (!id1.Exists && !id2.Exists) || id1.Equals(id2);
+
+#if DEBUG
+                static void verifyRemainderNonExistant(VariableIdentifier[] variablesBySlot, int startIndex)
+                {
+                    for (int i = startIndex; i < variablesBySlot.Length; i++)
+                    {
+                        Debug.Assert(!variablesBySlot[i].Exists);
+                    }
+                }
+#endif
+            }
+
+            internal void SnapshotWalker(BoundNode node)
+            {
+                if (_snapshotMap != null)
+                {
+                    _snapshotMap[node] = Snapshot.SnapshotWalker(_walkerToSnapshot);
```

Comment:
> _snapshotMap[node] = Snapshot.SnapshotWalker(_walkerToSnapshot); [](start = 19, length = 65)

Is there a way to assert that we are not overwriting values during the same pass?